docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -0,0 +1,436 @@
+common_1_ecall(enclave_foo);
+
+// ECALL to the enclave `bar`.
+common_1_ecall(enclave_bar);
+```
+
+Since the host can have only one implementation of `ecall_common_1_ecall` (the wrapper function),
+both invocations end up using the same function id, say `0`. However, this causes the mismatch
+on the call into the enclave `bar` where the expected id should be `2`.
+
+**Case Study 2: Selectively importing the ECALLs from same EDLs**
+
+Another scenario that both two enclaves selectively import ECALLs from the same EDLs but the
+selections are different.
+
+```
+// foo.edl
+enclave
+{
+    from "common_2.edl" import *;
+
+    trusted
+    {
+        public void foo_ecall();
+    };
+}
+```
+
+```
+// bar.edl
+enclave
+{
+    from "common_2.edl" import common_2_ecall_2;
+
+    trusted
+    {
+        public void bar_ecall();
+    };
+}
+```
+
+The resulting tables will be as follows.
+- `foo` ECALL table
+  ```
+  oe_ecall_func_t __oe_ecalls_table[] = {
+      (oe_ecall_func_t) ecall_common_2_ecall_1,
+      (oe_ecall_func_t) ecall_common_2_ecall_2,
+      (oe_ecall_func_t) ecall_foo_ecall
+  };
+  ```
+- `bar` ECALL table
+  ```
+  oe_ecall_func_t __oe_ecalls_table[] = {
+      (oe_ecall_func_t) ecall_common_2_ecall_2,
+      (oe_ecall_func_t) ecall_bar_ecall
+  };
+  ```
+
+Clearly, we can see the conflict on the function id.
+
+Proposed revision
+-----------
+
+The main goal of this proposal is to address the limitation that different enclaves on the
+same host cannot import the same EDL files (i.e., using the same set of ECALLs).
+To this end, this document proposes to refactor the ECALL calling mechanism as follows.
+
+**Maintain Local and Global ECALL tables on the Host**
+
+- Local Tables
+  Given that the rule of importing `EDL` files is flexible, the resulting
+  ECALL table is not deterministic; i.e., the same ECALL may have different ids
+  across enclaves. To solve the problem, this document uses an idea of storing the
+  list of ECALL id per enclave, which we refer to as local ECALL tables.
+  The local table is declared as follows.
+
+  ```
+  typedef struct _oe_ecall_id_t
+  {
+     uint64_t id;
+  } oe_ecall_id_t;
+
+  typedef struct _oe_enclave
+  {
+     ...
+     /* Per-edl ecall id table. */
+     oe_ecall_id_t ecall_id_table[OE_ECALL_MAX];
+     uint32_t ecall_id_table_size;
+     ...
+  } oe_enclave_t;
+  ```
+
+  The local table allows the host to retrieve the ECALL id per enclave instead
+  of using the hard-coded id in the same ECALL across different enclaves.
+
+- Global Table
+  To guarantee the constant-time look-up, the host needs to use the same id to retrieve
+  the ECALL id from the local table. To do so, the host additionally maintains a global
+  table that keeps such ids to which we refer as global ids. The detailed design is as follows.
+
+  - The global table is defined internally in the `oehost` as follows. The table stores name
+    of each ECALL and the index to each name is used as the global id.
```

Comment:
Why do you need both an index and a name?   Seems like you should be able to use one or the other without needing both.  You just need one predictable (not dynamic) global identifier (which could just be the name itself) that you can use as a key to look up the local ID in the host's per-enclave table.  That global identifier can be hard-coded be in the edger8r generated code.  And you don't need a global table then either, at least that I understand so far.

Reply:
The design is that the host stores the name only with the array
`const char* _ecall_table[OE_ECALL_MAX];`
and the index of each name in the array is used as global indices. 

Although the name (or any hard-coded global identifier) is sufficient to loop up the local id, we cannot guarantee the constant-time operations. Besides, there is no good data structures (e.g., hash map) in C unless we implement one.

The purpose of the global table is therefore keeping things simple: Keeping track of which ecall has been registered across multiple enclaves. Therefore, the host can ensure that the ecall with the same identifier will have the same index in the local table.

This can be illustrated by the example in this proposal:
	- `_ecall_table`
  ```
  _ecall_table[0]: { "ecall_common_1_ecall" }
  _ecall_table[1]: { "ecall_common_2_ecall_1" }
  _ecall_table[2]: { "ecall_common_2_ecall_2" }
  _ecall_table[3]: { "ecall_foo_ecall" }
  _ecall_table[4]: { "ecall_bar_ecall" }
  ```
- `foo` (at the time `bar` is not initialized yet)
  ```
  enclave->ecall_id_table[0]: { 0 /* ecall_common_1_ecall */ }
  enclave->ecall_id_table[1]: { 1 /* ecall_common_2_ecall_1 */ }
  enclave->ecall_id_table[2]: { 2 /* ecall_common_2_ecall_2 */ }
  enclave->ecall_id_table[3]: { 3 /* ecall_foo_ecall */ }
  ```
- `bar`
  ```
  enclave->ecall_id_table[0]: { 2 /* ecall_common_1_ecall */ }
  enclave->ecall_id_table[1]: { 0 /* ecall_common_2_ecall_1 */ }
  enclave->ecall_id_table[2]: { 1 /* ecall_common_2_ecall_2 */ }
  enclave->ecall_id_table[3]: { OE_ECALL_ID_NULL /* ecall_foo_ecall */ }
  enclave->ecall_id_table[4]: { 3 /* ecall_bar_ecall */ }
  ```

When host initializes the local table of `bar`, it basically replicates the global table such that the same global index can be used to locate the same ecall on different local tables. For example, the index of `ecall_common_1_ecall` is `0`. The global index `0` can be used to look up the local id on both `foo` (get `0`) and `bar` (get `2`).